front/components/ContentNodeTree.tsx

```diff
@@ -0,0 +1,352 @@
+};
+
+const useContentNodeTreeContext = () => {
+  const context = useContext(ContentNodeTreeContext);
+  if (!context) {
+    throw new Error(
+      "useContentNodeTreeContext must be used within a ContentNodeTreeContext"
+    );
+  }
+  return context;
+};
+
+interface ContentNodeTreeChildrenProps {
+  isRoundedBackground?: boolean;
+  isSearchEnabled?: boolean;
+  parentId: string | null;
+  parentIds: string[];
+  parentIsSelected?: boolean;
+}
+
+function ContentNodeTreeChildren({
+  isRoundedBackground,
+  isSearchEnabled,
+  parentId,
+  parentIds,
+  parentIsSelected,
+}: ContentNodeTreeChildrenProps) {
+  const { onDocumentViewClick, selectedNodes, setSelectedNodes } =
+    useContentNodeTreeContext();
+
+  const [search, setSearch] = useState("");
+  // This is to control when to dislpay the "Select All" vs "unselect All" button.
+  // If the user pressed "select all", we want to display "unselect all" and vice versa.
+  // But if the user types in the search bar, we want to reset the button to "select all".
+  const [selectAllClicked, setSelectAllClicked] = useState(false);
+
+  const { useResourcesHook } = useContentNodeTreeContext();
+
+  const { resources, isResourcesLoading, isResourcesError } =
+    useResourcesHook(parentId);
+
+  const filteredNodes = resources.filter(
+    (n) => search.trim().length === 0 || n.title.includes(search)
+  );
+
+  const getCheckedState = useCallback(
+    (node: BaseContentNode) => {
+      if (!selectedNodes) {
+        return "unchecked";
+      }
+
+      // If the parent is selected, the node is considered selected.
+      if (parentIsSelected) {
+        return "checked";
+      }
+
+      // Check if there is a local state for this node.
+      const localState = selectedNodes[node.internalId];
+      if (localState?.isSelected) {
+        return "checked";
+      }
+
+      const internalPartiallySelectedId = Object.values(selectedNodes)
+        .map((status) => status.parents)
+        .flat();
+      if (internalPartiallySelectedId.includes(node.internalId)) {
+        return "partial";
+      }
+
+      // Return false if no custom function is provided.
+      return "unchecked";
+    },
+    [parentIsSelected, selectedNodes]
+  );
+
+  if (isResourcesError) {
+    return (
+      <div className="text-warning text-sm">
+        Failed to retrieve permissions likely due to a revoked authorization.
+      </div>
+    );
+  }
+
+  const tree = (
+    <Tree isLoading={isResourcesLoading}>
+      {filteredNodes.map((n, i) => {
+        const checkedState = getCheckedState(n);
+
+        return (
+          <Tree.Item
+            key={n.internalId}
+            type={n.expandable ? "node" : "leaf"}
+            label={n.title}
+            visual={getVisualForContentNode(n)}
+            className="whitespace-nowrap"
+            checkbox={
+              n.preventSelection !== true && selectedNodes
+                ? {
+                    disabled: parentIsSelected || !setSelectedNodes,
```

Comment:
Might worth a small comment to explain that `setSelectedNodes` is not set while selection is first loaded.